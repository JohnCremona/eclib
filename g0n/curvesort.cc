// FILE CURVESORT.CC:  isogeny class id codes etc
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2007 John Cremona
// 
// This file is part of the mwrank/g0n package.
// 
// mwrank is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// mwrank is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with mwrank; if not, write to the Free Software Foundation,
// Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
// 
//////////////////////////////////////////////////////////////////////////

#define USE_NEW_CODE_LETTERS

int BA[] = {1,0};
int ACB[] = {0,2,1};
int BAC[] = {1,0,2};
int BCA[] = {1,2,0};
int CAB[] = {2,0,1};
int CBA[] = {2,1,0};
int ABDC[] = {0,1,3,2};
int ACBD[] = {0,2,1,3};
int ACDB[] = {0,2,3,1};
int ADBC[] = {0,3,1,2};
int ADCB[] = {0,3,2,1};
int BACD[] = {1,0,2,3};
int BADC[] = {1,0,3,2};
int BCAD[] = {1,2,0,3};
int BCDA[] = {1,2,3,0};
int BDAC[] = {1,3,0,2};
int CABD[] = {2,0,1,3};
int CBAD[] = {2,1,0,3};
int CDAB[] = {2,3,0,1};
int CDBA[] = {2,3,1,0};
int DABC[] = {3,0,1,2};
int DACB[] = {3,0,2,1};
int DBAC[] = {3,1,0,2};
int DCAB[] = {3,2,0,1};
int AEBDC[] = {0,4,1,3,2};
int AECBD[] = {0,4,2,1,3};
int AEDCB[] = {0,4,3,2,1};
int BCADE[] = {1,2,0,3,4};
int BCAED[] = {1,2,0,4,3};
int BEACD[] = {1,4,0,2,3};
int BECAD[] = {1,4,2,0,3};
int CEADB[] = {2,4,0,3,1};
int CABDE[] = {2,0,1,3,4};
int CEDAB[] = {2,4,3,0,1};
int CEDBA[] = {2,4,3,1,0};
int DAECB[] = {3,0,4,2,1};
int DCAEB[] = {3,2,0,4,1};
int DBCAE[] = {3,1,2,0,4};
int EABDC[] = {4,0,1,3,2};
int EACBD[] = {4,0,2,1,3};
int EBCAD[] = {4,1,2,0,3};
int EDACB[] = {4,3,0,2,1};
int ADECFB[] = {0,3,4,2,5,1};
int AECDBF[] = {0,4,2,3,1,5};
int BADECF[] = {1,0,3,4,2,5};
int BDAECF[] = {1,3,0,4,2,5};
int BFADCE[] = {1,5,0,3,2,4};
int CABEDF[] = {2,0,1,4,3,5};
int CAFBDE[] = {2,0,5,1,3,4};
int CEABDF[] = {2,4,0,1,3,5};
int DABFEC[] = {3,0,1,5,4,2};
int DCABEF[] = {3,2,0,1,4,5};
int EABCDF[] = {4,0,1,2,3,5};
int EBACDF[] = {4,1,0,2,3,5};
int ECABDF[] = {4,2,0,1,3,5};
int AEDBGFC[] = {0,4,3,1,6,5,2};
int AFGDEBC[] = {0,5,6,3,4,1,2};
int ECFBDGA[] = {4,2,5,1,3,6,0};
int EFCAGDB[] = {4,5,2,0,6,3,1};
int EFGCABD[] = {4,5,6,2,0,1,3};
int FGACBED[] = {5,6,0,2,1,4,3};
int FGBDACE[] = {5,6,1,3,0,2,4};
int FGDBAEC[] = {5,6,3,1,0,4,2};
int FGDBEAC[] = {5,6,3,1,4,0,2};
int FGDEABC[] = {5,6,3,4,0,1,2};
int AGBDEFHC[] = {0,6,1,3,4,5,7,2};
int AFHCGDEB[] = {0,5,7,2,6,3,4,1};
int BECADFGH[] = {1,4,2,0,3,5,6,7};
int EGBAFHCD[] = {4,6,1,0,5,7,2,3};
int GBEAHDFC[] = {6,1,4,0,7,3,5,2};

int booknumber0(int level, int form)  // utes number starting from 0
{
  if(level<56) return form;
  if(level>450) return form;
switch (level) {
case 56: return BA[form]; break;
case 77: return ACB[form]; break;
case 84: return BA[form]; break;
case 99: return ACBD[form]; break;
case 102: return ACB[form]; break;
case 106: return DACB[form]; break;
case 110: return CBA[form]; break;
case 114: return CAB[form]; break;
case 116: return CAB[form]; break;
case 118: return ACDB[form]; break;
case 120: return BA[form]; break;
case 121: return CABD[form]; break;
case 123: return BA[form]; break;
case 124: return BA[form]; break;
case 126: return BA[form]; break;
case 128: return ADCB[form]; break;
case 130: return ACB[form]; break;
case 132: return BA[form]; break;
case 136: return BA[form]; break;
case 140: return BA[form]; break;
case 141: return EBCAD[form]; break;
case 142: return EABDC[form]; break;
case 144: return BA[form]; break;
case 147: return ACB[form]; break;
case 150: return CAB[form]; break;
case 153: return ADCB[form]; break;
case 154: return ACB[form]; break;
case 155: return CBA[form]; break;
case 158: return EACBD[form]; break;
case 162: return ACBD[form]; break;
case 168: return BA[form]; break;
case 170: return DAECB[form]; break;
case 171: return ADBC[form]; break;
case 174: return CEDBA[form]; break;
case 175: return CAB[form]; break;
case 178: return BA[form]; break;
case 182: return CEADB[form]; break;
case 184: return DABC[form]; break;
case 185: return BCA[form]; break;
case 186: return ACB[form]; break;
case 187: return BA[form]; break;
case 189: return ADBC[form]; break;
case 190: return BAC[form]; break;
case 192: return ACBD[form]; break;
case 195: return CDBA[form]; break;
case 196: return BA[form]; break;
case 198: return CEDAB[form]; break;
case 200: return BECAD[form]; break;
case 201: return BCA[form]; break;
case 203: return BCA[form]; break;
case 205: return CAB[form]; break;
case 208: return DABC[form]; break;
case 210: return DBCAE[form]; break;
case 212: return BA[form]; break;
case 214: return DBAC[form]; break;
case 219: return BCA[form]; break;
case 221: return BA[form]; break;
case 222: return EDACB[form]; break;
case 234: return BCADE[form]; break;
case 235: return CAB[form]; break;
case 236: return BA[form]; break;
case 238: return DAECB[form]; break;
case 240: return BCAD[form]; break;
case 242: return BA[form]; break;
case 245: return CAB[form]; break;
case 246: return EFCAGDB[form]; break;
case 249: return BA[form]; break;
case 252: return BA[form]; break;
case 254: return DCAB[form]; break;
case 256: return BACD[form]; break;
case 262: return BA[form]; break;
case 264: return DABC[form]; break;
case 267: return BA[form]; break;
case 270: return BCDA[form]; break;
case 272: return ADBC[form]; break;
case 274: return CBA[form]; break;
case 278: return BA[form]; break;
case 285: return CAB[form]; break;
case 286: return BDAECF[form]; break;
case 288: return AECBD[form]; break;
case 291: return DBAC[form]; break;
case 294: return EFGCABD[form]; break;
case 297: return DACB[form]; break;
case 298: return BA[form]; break;
case 300: return ACDB[form]; break;
case 302: return BAC[form]; break;
case 304: return ECABDF[form]; break;
case 306: return CBAD[form]; break;
case 312: return CAFBDE[form]; break;
case 315: return BA[form]; break;
case 318: return DCAEB[form]; break;
case 320: return EABCDF[form]; break;
case 322: return BADC[form]; break;
case 324: return ABDC[form]; break;
case 325: return EABDC[form]; break;
case 326: return ACB[form]; break;
case 330: return AEDCB[form]; break;
case 333: return DABC[form]; break;
case 336: return CABEDF[form]; break;
case 338: return ADECFB[form]; break;
case 339: return BAC[form]; break;
case 342: return FGDEABC[form]; break;
case 345: return AECDBF[form]; break;
case 348: return BDAC[form]; break;
case 350: return BFADCE[form]; break;
case 352: return CEABDF[form]; break;
case 354: return DCABEF[form]; break;
case 360: return BCAED[form]; break;
case 364: return BA[form]; break;
case 366: return FGBDACE[form]; break;
case 368: return AEDBGFC[form]; break;
case 369: return BA[form]; break;
case 370: return ACBD[form]; break;
case 372: return ACDB[form]; break;
case 378: return GBEAHDFC[form]; break;
case 380: return BA[form]; break;
case 381: return BA[form]; break;
case 384: return BECADFGH[form]; break;
case 387: return CABDE[form]; break;
case 390: return AFGDEBC[form]; break;
case 392: return EBACDF[form]; break;
case 396: return BAC[form]; break;
case 400: return AFHCGDEB[form]; break;
case 402: return ABDC[form]; break;
case 404: return BA[form]; break;
case 405: return DABFEC[form]; break;
case 406: return ACDB[form]; break;
case 408: return ADCB[form]; break;
case 410: return BDAC[form]; break;
case 414: return CDAB[form]; break;
case 418: return BAC[form]; break;
case 423: return ECFBDGA[form]; break;
case 425: return ADBC[form]; break;
case 426: return CAB[form]; break;
case 427: return BCA[form]; break;
case 432: return EGBAFHCD[form]; break;
case 434: return AEBDC[form]; break;
case 435: return BACD[form]; break;
case 437: return BA[form]; break;
case 438: return FGACBED[form]; break;
case 440: return ADBC[form]; break;
case 441: return BADECF[form]; break;
case 442: return BEACD[form]; break;
case 446: return ADCB[form]; break;
case 448: return AGBDEFHC[form]; break;
case 450: return FGDBEAC[form]; break;

 default: return form;
}
  return form;  //default case for levels not yet sorted manually
}


const char alphabet[] = "abcdefghijklmnopqrstuvwxyz";

int booknumber(int level, int form)  // permutes number starting from 1
{
  return 1+booknumber0(level,form-1);
}

// Old codes were:  A,B,...,Z,AA,BB,...,ZZ,AAA,BBB,... etc

// Function to convert old code to integer (from 0)

int old_codeletter_to_int(char* code)  // i counts from 0!
{
  int i = code[0]-'A';
  int n=1; while(code[n]!='\0') n++;
  return 26*(n-1)+i;
}

// Function to convert integer (from 0) to old code

void old_codeletter(int i, char* code, int width=0)
{
  int n=width;    // pads string to this width with blanks
  code[n]='\0';
  while (n) code[--n]=' ';

  int nc = i%26;
  char c = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"[nc];
  n = 1 + (i-nc)/26;
  if(width==0) code[n]='\0';
  while (n) code[--n]=c;
  /*
  int j = old_codeletter_to_int(code);
  if(i==j) return;
  cout<<i<<" -> "<<code<<" -> "<<j<<endl;
  */
}

// old-new codes were:  a,b,...,z,aa,ab,...,az,ba,bb,... etc

// Function to convert old-new code to integer (from 0)

int old_new_codeletter_to_int(char* code)  // i counts from 0!
{
  int b = code[0]-'a';
  if(code[1]=='\0')  return b;
  int a = code[1]-'a';
  return 26*(b+1)+a;
}

// Function to convert integer (from 0) to old-new code

void old_new_codeletter(int i, char* code)  // i counts from 0!
{
  int b = i%26;
  int a = (i-b)/26;
  if (a==0) {code[0]=alphabet[b]; code[1]='\0';}
  else {code[0]=alphabet[a-1]; code[1]=alphabet[b]; code[2]='\0';}
  /*
  int j = codeletter_to_int(code);
  if(i==j) return;
  cout<<i<<" -> "<<code<<" -> "<<j<<endl;
  */
}

// new-new codes (from 01.08.05) are:  a,b,...,z,ba,bb,...,bz,ca,cb,... etc.,  i.e. straight base 26 with digits a=0, b=1, ..., z=25

// Function to convert new code to integer (from 0)

int codeletter_to_int(char* code)  // i counts from 0!
{
  int b = code[0]-'a';
  if(code[1]=='\0')  return b;
  int a = code[1]-'a';
  return 26*b+a;
}

// Function to convert integer (from 0) to new code

void new_codeletter(int i, char* code)  // i counts from 0!
{
  int b = i%26;
  int a = (i-b)/26;
  if (a==0) {code[0]=alphabet[b]; code[1]='\0';}
  else {code[0]=alphabet[a]; code[1]=alphabet[b]; code[2]='\0';}
  /*
  int j = codeletter_to_int(code);
  if(i==j) return;
  cout<<i<<" -> "<<code<<" -> "<<j<<endl;
  */
}

#ifdef USE_NEW_CODE_LETTERS
inline void codeletter(int i, char* code) {return new_codeletter(i,code);}
#else
inline void codeletter(int i, char* code) {return old_codeletter(i,code);}
#endif

