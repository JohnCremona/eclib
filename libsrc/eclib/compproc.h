// compproc.h: declarations of functions using complex numbers
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
#ifndef _ECLIB_COMPPROC_H_
#define _ECLIB_COMPPROC_H_

#include <eclib/interface.h>

int is_small(bigfloat x);
int is_small(const bigcomplex& z);
int is_real(const bigcomplex& z);

void orderreal(bigfloat& e1, bigfloat& e2, bigfloat& e3);  // puts in decreasing order
bigcomplex root(const bigcomplex& z, int n);
bigcomplex cagm(const bigcomplex& a, const bigcomplex& b);
bigcomplex normalize(bigcomplex& w1, bigcomplex& w2);
void getc4c6(const bigcomplex& w1, const bigcomplex& w2, bigcomplex& c4, bigcomplex &c6);
bigcomplex discriminant(const bigcomplex& b, const bigcomplex& c, const bigcomplex& d);

vector<bigcomplex> solvecubic(const bigcomplex& c1, const bigcomplex& c2, const bigcomplex& c3);
inline vector<bigcomplex> solvecubic(const bigfloat& c1, const bigfloat& c2, const bigfloat& c3)
{return solvecubic(bigcomplex(c1), bigcomplex(c2), bigcomplex(c3));}

vector<bigcomplex> solvequartic(const bigcomplex& a, const bigcomplex& b, const bigcomplex& c, const bigcomplex& d);
inline vector<bigcomplex> solvequartic(const bigfloat& a, const bigfloat& b, const bigfloat& c, const bigfloat& d)
{return solvequartic(bigcomplex(a), bigcomplex(b), bigcomplex(c), bigcomplex(d));}

vector<bigcomplex> solverealquartic(const bigfloat& a, const bigfloat& b, const bigfloat& c, const bigfloat& d, const bigfloat& e);

void quadsolve(const bigfloat& p, const bigfloat& q, bigcomplex& root1,bigcomplex& root2);

#endif
