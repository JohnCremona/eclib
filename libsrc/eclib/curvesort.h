// FILE CURVESORT.H:  isogeny class id codes etc
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2012 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////

#ifndef _ECLIB_CURVESORT_H
#define _ECLIB_CURVESORT_H      1
                           //flags that this file has been included

#include <string>
using namespace std;

#define USE_NEW_CODE_LETTERS


int booknumber0(int level, int form);  // permutes numbers starting from 0

const char alphabet[] = "abcdefghijklmnopqrstuvwxyz";

int booknumber(int level, int form);  // permutes numbers starting from 1

// new-new codes (from 01.08.05) are:
// a,b,...,z,ba,bb,...,bz,ca,cb,... etc., i.e. straight base 26 with
// digits a=0, b=1, ..., z=25

// Function to convert new code to integer (from 0) for any length of code.
// NB N=176400 has 516 < 26^2 newforms, with codes from a to vt!
int codeletter_to_int(string code);  // i counts from 0!

// Function to convert integer (from 0) to new code

string new_codeletter(int i);  // i counts from 0!

#ifdef USE_NEW_CODE_LETTERS
inline string codeletter(int i) {return new_codeletter(i);}
#else
inline string codeletter(int i) {return old_codeletter(i);}
#endif

#endif
