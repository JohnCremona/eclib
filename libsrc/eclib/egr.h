// egr.h : declaration of functions for reduction of points & component groups
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
// allow for multiple includes
#ifndef _ECLIB_EGR_H_
#define _ECLIB_EGR_H_

#include "curve.h"

// class to handle component groups at primes of bad reduction etc

// NB all of the internal methods assume that the curve is a global
// minimal model.  The normal use of the class will be via the
// function egr_index() which handles computation of a minimal model
// and transferring the points to the minimal model where necessary,

class ComponentGroups : public CurveRed {
public:
  explicit ComponentGroups(const Curvedata& CD) : CurveRed(CD) {;}
  explicit ComponentGroups(const CurveRed& CR) : CurveRed(CR) {;}
  //  ComponentGroups(const Curve& C) : CurveRed(C) {;}
  ComponentGroups() : CurveRed() {;}

// return 1 iff P mod p is nonsingular for all p (including infinity);
// else return 0 and put the first prime of bad reduction into p0:
  int HasGoodReduction(const Point& P, ZZ& p0) const
  {
    return P.has_good_reduction(the_bad_primes, p0, 1);
  }

// Returns [m] for cyclic of order m, [2,2] for 2*2 (type I*m, m even)
  vector<int> ComponentGroup(const ZZ& p) const;

// Returns 1 iff P and Q have same image in the component group at p:
  int InSameComponent(const Point& P, const Point& Q, const ZZ& p) const;

// For reduction type Im, multiplicative reduction where component
// group is cyclic of order m, returns a such that P mod p maps to a
// mod m in the component group 

// N.B.1 This will always return a in the range [0..m/2] and CANNOT
// distinguish between P and -P!  Be warned!

// N.B.2 The case of noncyclic component group is not handled here.

  long ImageInComponentGroup(const Point&P, const ZZ& p, vector<int> grp) const;
  long ImageInComponentGroup_Im(const Point&P, const ZZ& p, int m) const;
  long ImageInComponentGroup_Im_pm(const Point&P, const ZZ& p, int m) const;

// Return least j>0 such that j*P has good reduction at p, i.e. the
// order of the image of P in the component group at p; the component
// group is given
  int OrderInComponentGroup(const Point& P, const ZZ& p, const vector<int>& grp);

// replace (independent) points in Plist with a new set which spans
// the subgroup of the original with good reduction at p, returning
// the index
  int gr1prime(vector<Point>& Plist, const ZZ& p);

// replace (independent) points in Plist with a new set which spans
// the subgroup of the original with good reduction at all p in plist,
// returning the overall index
  int grprimes(vector<Point>& Plist, const vector<ZZ>& plist);

// replace (independent) points in Plist with a new set which spans
// the subgroup of the original with good reduction at all p,
// returning the overall index
  int grprimes(vector<Point>& Plist) {return grprimes(Plist,the_bad_primes);}

// replaces the (independent) points with a new set which spans the
// subgroup of the original with good reduction at all p,
// returning the overall index
  int egr_subgroup(vector<Point>& Plist, int real_too=1);
};

// returns the index in the subgroup generated by the given points of
// its egr subgroup (the points are unchanged)
ZZ egr_index(const vector<Point>& Plist, int real_too=1);

// Given a list of points P1,...,Pn and a prime p, this returns a
// vector [c1,c2,...,cn] where the image of Pi in the component group
// is ci mod m, where m is the exponent of the component group at p.
// 
// Each ci is a vector of length 1 or 2 (the latter for when the
// component group is C2xC2), not just an integer.
//
// If p=0 then m=1 or 2 (m=2 iff there are two real components and at
// least one P_i is not in the connected component)
//

vector<vector<int> >  MapPointsToComponentGroup(const ComponentGroups& CG, const vector<Point>& Plist,  const ZZ& p);

#endif
