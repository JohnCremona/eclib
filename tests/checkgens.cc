// FILE CHECKGENS.CC -- Program to check input gens are Mordell-Weil basis
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2007 John Cremona
// 
// This file is part of the mwrank/g0n package.
// 
// mwrank is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// mwrank is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with mwrank; if not, write to the Free Software Foundation,
// Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
// 
//////////////////////////////////////////////////////////////////////////


#include <fstream>
#include "points.h"
#include "mwprocs.h"
#include "sifter.h"
#include "compproc.h"
#include "egr.h"
#include "htconst.h"

#define INPUT_CLASS_IS_LETTER // we only use letters now!

Curve C;
Curvedata CD;
vector<Point> plist;
int verbose;
char genfile[30];
char ccode[5];

#include "curvesort.cc" // for codeletter() function

int main()
{  
//  cout<<"precision = "<<decimal_precision()<<endl;
  set_precision(15);
//  cout<<"precision = "<<decimal_precision()<<endl;
  cin.flags( cin.flags() | ios::dec );  //force decimal input (bug fix)

  int rank, rank2, i;
  long cond, ncurve;
  bigfloat reg, hmax;;
  cerr<<"verbose (0/1)? ";             cin >>verbose;
  verbose=0;
//  cerr<<"\nLimit on height of search (-1 to just check points are on curve)? ";
//  cin >> hmax;
  hmax=-1;
  cerr <<"\n";
  cerr << "input filename for curves and generators? "; 
  cin >> genfile;
  ifstream genin;
  genin.open(genfile);
  if(!genin.is_open()) {cerr<<"Unable to open file " << genfile << "\n"; abort();}
  cerr<<endl;

  while (!genin.eof())
    {
      genin >> cond;
      //      cout<<"input conductor="<<cond<<endl;
      if(cond==0) break;
#ifdef INPUT_CLASS_IS_LETTER
      genin >> ccode;
#else
      long nclass;
      genin >> nclass; 
      codeletter((nclass-1),ccode);
      //      cout<<"After input, nclass="<<nclass<<endl;
#endif
      genin >> ncurve;
      //      cout<<"After input, ncurve="<<ncurve<<endl;
      genin>>C;
      //      cout<<"After input, C="<<C<<endl;
      cout << cond<<ccode<<ncurve<<"\t";
      cout << C << ":\t";
      CD = Curvedata(C);
      if(verbose) cout <<endl;
      genin >> rank;
      cout << "r = "<<rank<<"\t";
      if(verbose) cout <<endl;
      if (rank==0) {cout<<" OK\n"; continue;}
      Point P(CD);
      plist.clear();
      while(plist.size()<rank)
	{
	  genin >> P; 
	  cout << P;
	  if(P.isvalid()) 
	    {
	      plist.push_back(P);
	      cout << " OK ";
	    }
	  else 
	    {
	      cout << " not on curve! ";  abort();
	    }
	}

      // Do nothing at all apart from checking points are on curve
      // (since we have a separate program for saturation now):

      if(hmax<0) {cout<<endl;continue;}

      bigfloat ht, maxht = to_bigfloat(0);
      for (i=0; i<rank; i++)
	{ 
	  ht  = height(plist[i]);
	  if(ht>maxht) maxht=ht;
	}

      mw mwbasis(&CD,verbose,1);
      mwbasis.process(plist);
      rank2 = mwbasis.getrank();
      int rank_under = (rank2<rank);
      if(rank_under)
	{
	  cout<<"\nInput points only have rank "<<rank2<<", not "<<rank<<"\n";
//	  continue;
	}
      else cout<<" INDEP ";
      // Now we check that the points are independent in E(Q)/2E(Q):

      long naux=rank;
      long r2=0;
      while((naux<100)&&(r2<rank))
	{
	  naux+=5;
	  sifter box(&CD, naux, verbose);
	  box.process(plist);
	  r2 = box.getrank();
	}
      if(r2==rank)
	{
	  cout<<" INDEP_MOD_2 ";
	  if(verbose)
	    cout<<"(Points are independent mod 2)"<<endl;
	}
      else
	cout<<"\nPoints may be dependent mod 2, r2 = "<<r2<<endl;



      if(hmax<1) {cout<<endl;}
      else
	{
      reg = mwbasis.regulator();
	
      if(verbose) cout<<"\nMax height = "<<maxht<<endl;
      if(rank==1) 
	if(r2==rank) maxht/=9;	else maxht/=4;  
      // since in this case we know that the index is at least 2 (or 3).
      double ht_bound = height_constant(CD);
      if(verbose) cout << "height bound = " << ht_bound << "\n";
      bigfloat hlim = maxht+ht_bound;
      if(verbose||(hlim>hmax)) 
	{
	  if(!verbose) cout<<"\n";
	  cout<<"Bound on naive height of extra generators = "<<hlim<<endl;
	}      
      if(hlim>hmax)
	{
	  cout<<"Only searching up to height "<<hmax<<endl;
	  hlim=hmax;
	}
      else if(!verbose) cout<<endl;
      
      sieve s(&CD, &mwbasis, 2, 0);
      s.search(hlim);

      rank2 = mwbasis.getrank();
      if(rank2!=rank)
	{
	  cout << "\nPoints found have rank " << rank2<<", not "<<rank<<"\n";
	  continue;
	}
      if(rank_under)
	cout << "New basis: "<<mwbasis.getbasis()<<endl;

      bigfloat reg2 = mwbasis.regulator();

      long index = I2long(Iround((sqrt(reg/reg2)+0.01)));

      if(index>1)
	{
	  if(!verbose) cout<<"\n";
	  cout << "Points found contain original with index " << index <<"\n";

	  vector<Point> b = mwbasis.getbasis();
	  for (i=0; i<rank; i++)
	    { Point P = b[i];
	      cout << "\nGenerator "<<(i+1)<<" is "<<P<<"; ";
	      cout << "height "<<height(P);
	    }
	  cout<<"\nNew regulator is "<<reg2<<" (old was "<<reg<<")"<<endl<<endl;
	}
	}
//      else cout << "\n";
    }
}         // end main()


