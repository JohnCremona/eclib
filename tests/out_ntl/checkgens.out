268605a1	[0,0,1,1032,2673]:	r = 1	torsion group structure =  [ ] (torsion rank 0).
P = [41:337:1] OK, order=oo, height=0.59990651734770452119201067743.
268605b1	[0,0,1,14802,-4767741]:	r = 1	torsion group structure =  [ ] (torsion rank 0).
P = [52199:344443:343] OK, order=oo, height=4.8009119417025658580313798627.
268605c1	[1,-1,0,-450,2781]:	r = 2	torsion group structure =  [ ] (torsion rank 0).
P = [-154:585:8] OK, order=oo, height=2.4620339650023437699163975371.
P = [76:597:1] OK, order=oo, height=1.9744004986562372890973350171.
268605d1	[1,-1,0,-1194,15983]:	r = 1	torsion group structure =  [ 2 ] (torsion rank 1).
P = [-314:437:8] OK, order=oo, height=4.2655328871552274826126272117.
P = [22:-11:1] OK, order=2, height=0.
268605d2	[1,-1,0,-2319,-17992]:	r = 1	torsion group structure =  [ 2 2 ] (torsion rank 2).
P = [-117782:149171:2744] OK, order=oo, height=8.5310657743104549652252544234.
P = [52:-26:1] OK, order=2, height=0.
P = [-8:4:1] OK, order=2, height=0.
268605d3	[1,-1,0,8256,-142777]:	r = 1	torsion group structure =  [ 2 ] (torsion rank 1).
P = [2451471778:11277201335:138991832] OK, order=oo, height=17.062131548620909930450508847.
P = [134:-67:8] OK, order=2, height=0.
268605d4	[1,-1,0,-30894,-2081107]:	r = 1	torsion group structure =  [ 2 ] (torsion rank 1).
P = [-64232978894:39611125679:622835864] OK, order=oo, height=17.062131548620909930450508847.
P = [-826:413:8] OK, order=2, height=0.
268605e1	[1,-1,0,8256,753083]:	r = 1	torsion group structure =  [ 2 ] (torsion rank 1).
P = [395144:10859243:512] OK, order=oo, height=8.5829151366781123251964064488.
P = [-62:31:1] OK, order=2, height=0.
268605e2	[1,-1,0,-91149,9441080]:	r = 1	torsion group structure =  [ 2 2 ] (torsion rank 2).
P = [317636:-1639318:1331] OK, order=oo, height=4.2914575683390561625982032244.
P = [998:-499:8] OK, order=2, height=0.
P = [220:-110:1] OK, order=2, height=0.
268605e3	[1,-1,0,-1413024,646849205]:	r = 1	torsion group structure =  [ 2 ] (torsion rank 1).
P = [7372618:-6306479:10648] OK, order=oo, height=8.5829151366781123251964064488.
P = [5510:-2755:8] OK, order=2, height=0.
268605e4	[1,-1,0,-359754,-73128097]:	r = 1	torsion group structure =  [ 2 ] (torsion rank 1).
P = [265770428:14415192911:85184] OK, order=oo, height=8.5829151366781123251964064488.
P = [-3514:1757:8] OK, order=2, height=0.
268605f1	[0,0,1,-110937,14226165]:	r = 1	torsion group structure =  [ ] (torsion rank 0).
P = [1554:631:8] OK, order=oo, height=1.9389206422900754768913015595.
268605g1	[0,0,1,-747,8255]:	r = 0	torsion group structure =  [ ] (torsion rank 0).
